cpp/src/Ice/ConnectionI.cpp

```diff
@@ -2283,13 +2283,13 @@ Ice::ConnectionI::idleCheck(
     const chrono::seconds& idleTimeout) noexcept
 {
     std::lock_guard lock(_mutex);
-    if (_state == StateActive || _state == StateHolding)
+    // When _timer->isScheduled(idleCheckTimerTask) returns true, it means a read rescheduled the
+    // timer task while we were waiting to lock _mutex. We don't do anything in this case.
+    if ((_state == StateActive || _state == StateHolding) && !_timer->isScheduled(idleCheckTimerTask))
     {
-        // _timer->cancel(task) returns true if a concurrent read rescheduled the timer task.
-        if (_transceiver->isWaitingToBeRead() || _timer->cancel(idleCheckTimerTask))
+        if (_transceiver->isWaitingToBeRead())
         {
-            // Schedule or reschedule timer task. Reschedule in the rare case where a concurrent read scheduled the task
-            // already.
+            // Schedule timer task.
```

Comment:
It's possible we use could schedule instead of reschedule here.

It depends if the decorator only (re)schedules this timer task while ConnectionI's mutex is locked.